dot/parachain/prospective-parachains/fragment-chain/fragment_chain.go

```diff
@@ -0,0 +1,1196 @@
+			},
+			state: entry.state,
+		}
+	}
+
+	return clone
+}
+
+func NewCandidateStorage() *CandidateStorage {
+	return &CandidateStorage{
+		byParentHead:    make(map[common.Hash]map[parachaintypes.CandidateHash]any),
+		byOutputHead:    make(map[common.Hash]map[parachaintypes.CandidateHash]any),
+		byCandidateHash: make(map[parachaintypes.CandidateHash]*CandidateEntry),
+	}
+}
+
+func (c *CandidateStorage) AddPendingAvailabilityCandidate(
+	candidateHash parachaintypes.CandidateHash,
+	candidate parachaintypes.CommittedCandidateReceipt,
+	persistedValidationData parachaintypes.PersistedValidationData,
+) error {
+	entry, err := NewCandidateEntry(candidateHash, candidate, persistedValidationData, Backed)
+	if err != nil {
+		return err
+	}
+
+	return c.addCandidateEntry(entry)
+}
+
+// Len return the number of stored candidate
+func (c *CandidateStorage) Len() int {
+	return len(c.byCandidateHash)
+}
+
+func (c *CandidateStorage) addCandidateEntry(candidate *CandidateEntry) error {
+	_, ok := c.byCandidateHash[candidate.candidateHash]
+	if ok {
+		return ErrCandidateAlradyKnown
+	}
+
+	// updates the reference parent hash -> candidate
+	// we don't check the `ok` value since the key can
+	// exists in the map but pointing to a nil hashset
+	setOfCandidates := c.byParentHead[candidate.parentHeadDataHash]
+	if setOfCandidates == nil {
+		setOfCandidates = make(map[parachaintypes.CandidateHash]any)
+	}
+	setOfCandidates[candidate.candidateHash] = struct{}{}
+	c.byParentHead[candidate.parentHeadDataHash] = setOfCandidates
+
+	// udpates the reference output hash -> candidate
+	setOfCandidates = c.byOutputHead[candidate.outputHeadDataHash]
+	if setOfCandidates == nil {
+		setOfCandidates = make(map[parachaintypes.CandidateHash]any)
+	}
+	setOfCandidates[candidate.candidateHash] = struct{}{}
+	c.byOutputHead[candidate.outputHeadDataHash] = setOfCandidates
+
+	c.byCandidateHash[candidate.candidateHash] = candidate
+	return nil
+}
+
+func (c *CandidateStorage) removeCandidate(candidateHash parachaintypes.CandidateHash) {
+	entry, ok := c.byCandidateHash[candidateHash]
+	if !ok {
+		return
+	}
+
+	delete(c.byCandidateHash, candidateHash)
+
+	if setOfCandidates, ok := c.byParentHead[entry.parentHeadDataHash]; ok {
+		delete(setOfCandidates, candidateHash)
+		if len(setOfCandidates) == 0 {
+			delete(c.byParentHead, entry.parentHeadDataHash)
+		}
+	}
+
+	if setOfCandidates, ok := c.byOutputHead[entry.outputHeadDataHash]; ok {
+		delete(setOfCandidates, candidateHash)
+		if len(setOfCandidates) == 0 {
+			delete(c.byOutputHead, entry.outputHeadDataHash)
+		}
+	}
+}
+
+func (c *CandidateStorage) markBacked(candidateHash parachaintypes.CandidateHash) {
+	entry, ok := c.byCandidateHash[candidateHash]
+	if !ok {
+		fmt.Println("candidate not found while marking as backed")
+	}
+
+	entry.state = Backed
+}
+
+func (c *CandidateStorage) contains(candidateHash parachaintypes.CandidateHash) bool {
+	_, ok := c.byCandidateHash[candidateHash]
+	return ok
+}
```

Comment:
Do we need this method? I would have instead used the map directly without declaring a method.